src/common/datasource/src/file_format/json.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::BufReader;
+use std::sync::Arc;
+
+use arrow::datatypes::SchemaRef;
+use arrow::json::reader::{infer_json_schema_from_iterator, ValueIter};
+use async_trait::async_trait;
+use object_store::ObjectStore;
+use snafu::ResultExt;
+
+use crate::compression::CompressionType;
+use crate::error::{self, Result};
+use crate::file_format::{self, FileFormat};
+use crate::util::io::SyncIoBridge;
+
+#[derive(Debug)]
+pub struct JsonFormat {
+    pub schema_infer_max_record: Option<usize>,
+    pub compression_type: CompressionType,
+}
+
+impl Default for JsonFormat {
+    fn default() -> Self {
+        Self {
+            schema_infer_max_record: Some(file_format::DEFAULT_SCHEMA_INFER_MAX_RECORD),
+            compression_type: CompressionType::UNCOMPRESSED,
+        }
+    }
+}
+
+#[async_trait]
+impl FileFormat for JsonFormat {
+    async fn infer_schema(&self, store: &ObjectStore, path: String) -> Result<SchemaRef> {
+        let reader = store
+            .reader(&path)
+            .await
+            .context(error::ReadObjectSnafu { path: &path })?;
+
+        let decoded = self.compression_type.convert_async_read(reader);
+
+        let mut reader = BufReader::new(SyncIoBridge::new(decoded));
+
+        let iter = ValueIter::new(&mut reader, self.schema_infer_max_record);
+
+        let schema = infer_json_schema_from_iterator(iter)
+            .context(error::InferSchemaSnafu { path: &path })?;
```

Comment:
`arrow_json` provides [method](https://docs.rs/arrow-json/36.0.0/arrow_json/struct.RawReader.html#method.schema) to get schema from reader. Can we make use of it?

This comment also applies to csv format

Reply:
No, we can't.
I looked at the source code, and `RawReaderBuilder` accepts a known `schema` as an argument at the `new` method.  
```rust
impl RawReaderBuilder {
    pub fn new(schema: SchemaRef) -> Self {
        Self {
            batch_size: 1024,
            coerce_primitive: false,
            schema,
        }
    }
``` 
The document also mentioned this point: 
> Reads JSON data with a known schema directly into arrow [RecordBatch](https://docs.rs/arrow-array/36.0.0/x86_64-unknown-linux-gnu/arrow_array/record_batch/struct.RecordBatch.html)